### PR TITLE
Add endpoints. Made clients to postprocess hadoop jobs

### PR DESCRIPTION
**Pull request by jerryli9876**
Made postprocessing framework, so that it is straightforward to add new endpoints. Made postprocessors which write to servers and files.

### CODE REVIEW

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/PerServiceClient.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package com.twitter.zipkin.hadoop
```
Comment (caniszczyk) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/PerServiceClient.scala:17]: Please add copyright headers to new source files.

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/PerServicePairClient.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.twitter.zipkin.hadoop
```
Comment (caniszczyk) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/PerServicePairClient.scala:17]: Please add copyright headers to new source files.

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToFileClient.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
+package com.twitter.zipkin.hadoop
```
Comment (caniszczyk) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToFileClient.scala:17]: Please add copyright headers to new source files.

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/ServiceNameList.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
+package com.twitter.zipkin.hadoop
```
Comment (caniszczyk) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/ServiceNameList.scala:17]: Please add copyright headers to new source files.

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToServerClient.scala [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package com.twitter.zipkin.hadoop
```
Comment (caniszczyk) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToServerClient.scala:17]: Please add copyright headers to new source files.

File: zipkin-hadoop-job-runner/src/test/scala/com/twitter/zipkin/hadoop/StandardizedServiceNameListSpec.scala [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package com.twitter.zipkin.hadoop
```
Comment (caniszczyk) [inline zipkin-hadoop-job-runner/src/test/scala/com/twitter/zipkin/hadoop/StandardizedServiceNameListSpec.scala:1]: Please add copyright headers to new source files.

**Comment by franklinhu** [comment]
Should make the title and description of the pull request a bit more descriptive

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/HadoopJobClient.scala [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Starts the postprocessing for the client
+   * @param filename the input filename
+   * @param output the output filename
+   */
+  def start(filename : String, output : String)
+
+  /**
+   * Processes a single file
```
Comment (johanoskarsson) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/HadoopJobClient.scala]: Please describe what the input format is like.

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/HadoopJobClient.scala [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+  protected val DELIMITER = ":"
+  val ssnm = if (combineSimilarNames) new StandardizedServiceNameList else new ServiceNameList
+
+  /**
+   * Populate the name list
+   * @param s Scanner representing the name list
+   */
+
+  def populateSsnm(s: Scanner)
```
Comment (johanoskarsson) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/HadoopJobClient.scala]: What does ssnm stand for? Worth expanding into the full name or at least describe in comment?

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToServerClient.scala [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      val protocol = new TBinaryProtocol(transport)
+      client = new gen.ZipkinCollector.Client(protocol)
+      transport.open()
+      // Read file
+      println("Connected to server...populating ssnm")
+      populateServiceNameList(new Scanner(new File(filename)))
+      println("Finished populating ssnm...beginning processFile")
+      processFile(new Scanner(new File(filename)))
+    } catch {
+      case se: SocketException => se.printStackTrace()
```
Comment (johanoskarsson) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToServerClient.scala]: Since you're doing the same for all of them you can just catch Throwable.

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToServerClient.scala [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      case tte : TTransportException => tte.printStackTrace()
```
Comment (johanoskarsson) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/WriteToServerClient.scala]: Worth looking into how much work it would be to use util-logging instead lf println:ing stuff. That way we can have the default to log to the console but can also redirect to local file or via scribe.

File: zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/sources/Util.scala [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+object Util {
+
+  /**
+   * Given two strings, finds the minimum edit distance between them given only substitutions, deletions, and additions with the
+   * Levenshtein algorithm.
+   * @param s1 a String
+   * @param s2 another String
+   * @return the edit distance between them after they're converted to lower case.
+   */
+  def getLevenshteinDistance(s1 : String, s2 : String) : Int = {
```
Comment (johanoskarsson) [inline zipkin-hadoop-job-runner/src/main/scala/com/twitter/zipkin/hadoop/sources/Util.scala]: Is there no library we rely on that has this already? Util? Apache Commons? Worth moving to util otherwise?

File: zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/FindDuplicateTraces.scala [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      if (first < Int.MaxValue && last > -1) Some(List(first, last)) else None
+    }.groupBy('trace_id){ _.reduce('first_and_last_timestamps -> 'first_and_last_timestamps) { (left : List[Long], right : List[Long]) =>
+        val first = if (left(0) > right(0)) right(0) else left(0)
+        val last = if (left(1) > right(1)) left(1) else right(1)
+        List(first, last)
+      }
+    }
+    .filter('first_and_last_timestamps) { timestamps : List[Long] =>
+      val dur = (timestamps(1) - timestamps(0)) / 1000000
+      dur >= 10 * 60
```
Comment (johanoskarsson) [inline zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/FindDuplicateTraces.scala]: Move this out to a constant or even a parameter?

File: zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/FindDuplicateTraces.scala [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+      }
+      if (first < Int.MaxValue && last > -1) Some(List(first, last)) else None
+    }.groupBy('trace_id){ _.reduce('first_and_last_timestamps -> 'first_and_last_timestamps) { (left : List[Long], right : List[Long]) =>
+        val first = if (left(0) > right(0)) right(0) else left(0)
+        val last = if (left(1) > right(1)) left(1) else right(1)
+        List(first, last)
+      }
+    }
+    .filter('first_and_last_timestamps) { timestamps : List[Long] =>
+      val dur = (timestamps(1) - timestamps(0)) / 1000000
```
Comment (johanoskarsson) [inline zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/FindDuplicateTraces.scala]: Rename dur to something that also tells us the unit type

File: zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/MemcacheRequest.scala [hunk 14]
```diff
@@ -29,26 +29,26 @@ class MemcacheRequest(args : Args) extends Job(args) with DefaultDateRangeJob {
...
-        }
+  val memcacheNames = preprocessed
+    .flatMap('binary_annotations -> 'memcacheNames){ bal : List[BinaryAnnotation] =>
         // from the binary annotations, find the value of the memcache visits if there are any
         var memcachedKeys : Option[BinaryAnnotation] = None
-        bl.foreach { ba : BinaryAnnotation => if (ba.key == "memcached.keys") memcachedKeys = Some(ba) }
-        for (cs <- clientSent; key <- memcachedKeys)
-          yield (cs.getHost.service_name, new String(Util.getArrayFromBuffer(key.value)))
-      }
+        bal.foreach { ba : BinaryAnnotation => if (ba.key == "memcached.keys") memcachedKeys = Some(ba) }
```
Comment (johanoskarsson) [inline zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/MemcacheRequest.scala]: You should be able to use find here instead of foreach.

File: zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/WorstRuntimesPerTrace.scala [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+  val preprocessed = PreprocessedSpanSource()
+    .read
+    .mapTo(0 -> ('service, 'trace_id, 'annotations)) {
+    s : SpanServiceName => (s.service_name, s.trace_id, s.annotations.toList)
+  }
+
+  val result = preprocessed
+    // let's find those client annotations and convert into service name and duration
+    .flatMap('annotations -> 'duration) { annotations: List[Annotation] =>
+    var clientSend: Option[Annotation] = None
```
Comment (johanoskarsson) [inline zipkin-hadoop/src/main/scala/com/twitter/zipkin/hadoop/WorstRuntimesPerTrace.scala]: Indentation is out of whack here and below

File: zipkin-hadoop/src/scripts/scald.rb [hunk 16]
```diff
@@ -32,7 +32,7 @@
 ZIPKIN_HADOOP_ROOT=File.expand_path(File.dirname(ORIGINAL_FILE)+"/../../")
 JARFILE=ZIPKIN_HADOOP_ROOT + "/target/zipkin-hadoop-assembly-#{JAR_VERSION}.jar" #what jar has all the depencies for this job
 puts JARFILE
-HOST="my.remote.host" #where the job is rsynced to and run
+HOST="hadoopnest1.smf1.twitter.com" #where the job is rsynced to and run
```
Comment (johanoskarsson) [inline zipkin-hadoop/src/scripts/scald.rb]: Revert this change

**Comment by johanoskarsson** [comment]
+1
